src/Orleans/Messaging/Message.cs

```diff
@@ -446,7 +446,10 @@ private static object DeserializeBody(List<ArraySegment<byte>> bytes)
 
         public Message()
         {
-            headers = new Dictionary<Header, object>();
+            // average headers items count is 14 items, and while the Header enum contains 18 entries
+            // the closest prime number is 17; assuming that possibility of all 18 headers being at the same time is low enough to
+            // choose 17 in order to avoid allocations of two additional items on each call, and allocate 37 instead of 19 in rare cases
+            headers = new Dictionary<Header, object>(17);
```

Comment:
I'm good with 16,17, or 32..  All should be better than what we have.  I don't have metrics on how big this dictionary gets and under what conditions, so tuning for the common case is reasonable.